crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -460,13 +460,12 @@ impl<'db> TypeInferenceBuilder<'db> {
             panic!("function type params scope without type params");
         };
 
-        // TODO: this should also be applied to parameter annotations.
         if !self.is_stub() {
             self.infer_optional_expression(function.returns.as_deref());
+            self.infer_parameters(&function.parameters);
```

Comment:
It's not correct to defer _all_ of parameter inference, only annotations (not defaults). So I think we can't just move `infer_parameters` in here, it'll need to get split up. We could just have `infer_parameter_defaults` and `infer_parameter_annotations`. Or we can call `infer_parameters` in both non-deferred and deferred cases, and do the `is_stub` checking down in `infer_parameter_with_default` and `infer_parameter`. Either could be OK, not sure which will work out nicer.